go/vt/wrangler/reparent.go

```diff
@@ -599,11 +607,34 @@ func (wr *Wrangler) plannedReparentShardLocked(ctx context.Context, ev *events.R
 		wr.logger.Infof("promote replica %v", masterElectTabletAliasStr)
 		event.DispatchUpdate(ev, "promoting replica")
 
+		waitCtx, waitCancel := context.WithTimeout(ctx, waitReplicasTimeout)
+		defer waitCancel()
+
+		err = wr.tmc.WaitForPosition(waitCtx, masterElectTabletInfo.Tablet, rp)
+
+		if err != nil || ctx.Err() != nil && ctx.Err() == context.DeadlineExceeded {
+			// If the new master fails to catch up within the timeout,
+			// we try to roll back to the original master before aborting.
+			// It is possible that we have used up the original context, or that
+			// not enough time is left on it before it times out.
+			// But at this point we really need to be able to Undo so as not to
+			// leave the cluster in a bad state.
+			// So we create a fresh context based on context.Background().
+			undoCtx, undoCancel := context.WithTimeout(context.Background(), *topo.RemoteOperationTimeout)
+			defer undoCancel()
+			if err1 := wr.tmc.UndoDemoteMaster(undoCtx, oldMasterTabletInfo.Tablet); err1 != nil {
+				log.Warningf("Encountered error %v while trying to undo DemoteMaster", err1)
```

Comment:
I prefer to print errors at the end since they can often be long:

```suggestion
				log.Warningf("Encountered error while trying to undo DemoteMaster: %v", err1)
```